zkevm-circuits/src/evm_circuit/execution/memory.rs

```diff
@@ -0,0 +1,378 @@
+use crate::{
+    evm_circuit::{
+        execution::{
+            bus_mapping_tmp::{Block, Call, ExecStep, Transaction},
+            ExecutionGadget,
+        },
+        param::MAX_GAS_SIZE_IN_BYTES,
+        step::ExecutionState,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{
+                ConstraintBuilder, StepStateTransition,
+                Transition::{Delta, To},
+            },
+            from_bytes,
+            math_gadget::IsEqualGadget,
+            memory_gadget::MemoryExpansionGadget,
+            select, MemoryAddress, Word,
+        },
+    },
+    util::Expr,
+};
+use bus_mapping::{eth_types::ToLittleEndian, evm::OpcodeId};
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+use std::convert::TryInto;
+
+#[derive(Clone, Debug)]
+pub(crate) struct MemoryGadget<F> {
+    same_context: SameContextGadget<F>,
+    address: MemoryAddress<F>,
+    value: Word<F>,
+    memory_expansion: MemoryExpansionGadget<F, MAX_GAS_SIZE_IN_BYTES>,
+    is_mload: IsEqualGadget<F>,
+    is_mstore8: IsEqualGadget<F>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for MemoryGadget<F> {
+    const NAME: &'static str = "MEMORY";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::MEMORY;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
```

Comment:
One question here. Why not put `opcode` into the `MemoryGadget` struct while everything else is in the struct? I also didn't the assignment to this `opcode` cell.

Reply:
This is done by `SameContextGadget` ([here](https://github.com/appliedzkp/zkevm-circuits/pull/196/files#diff-08d58ebc59ab6c11bf7c049e859441eed2319a03feb3fdffaddaba73f8a8c829R90-R91)), was intended to reduce the redudancy. Let me try to fix this ambiguity by move the opcode query into the `SameContextGadget`.

Reply:
Just realized why I left opcode in `ExecutionGadget`, because some of step state transition depends on the opcode, so we need to query the opcode first to generate such expression.

Reply:
I see.